Rest test style

oxygen0211 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Convert REST tests to pytest method style. Tests for 'binary_sensor', 'sensor' and 'notify' have already been converted for solving #41973 (kudos @bdraco), so not touching them.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #40885
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

oxygen0211 committed:
Convert REST switch setup tests to pytest method style

oxygen0211 committed:
Convert REST switch tests to pytest method style

oxygen0211 committed:
Refactor test switch code

Extract setup of test switch and templates into helper method, remove unneeded references to hass.loop

bdraco commented on tests/components/rest/test_switch.py:
```diff
@@ -16,254 +16,239 @@
+
+
+"""Tests for REST switch platform."""
+
+name = "foo"
+method = "post"
+resource = "http://localhost/"
+state_resource = resource
+headers = {"Content-type": CONTENT_TYPE_JSON}
+auth = None
```
Please make constants UPPERCASE and move to the top of the file.

oxygen0211 replied:
Done

bdraco reviewed (changes requested):
Ideally we setup the integration and test with the core apis instead of calling `rest.RestSwitch`

I realize that is a bigger ask so I think this is mergeable with just addressing the constants and the rest can be done in a new PR.

oxygen0211 committed:
Reformat constants to proper styling

oxygen0211 commented:
Hi @bdracto,
thanks for your review. I really much appreciate your statement about refactoring the tests. Not having such deep knowledge about the component and python and still learning in pytest (and some python specifica), I try to only make the minimal needed changes on these test style changes in order not to accidentally change the test semantics in an unwanted manor.

oxygen0211 commented:
Hi @bdraco ,

Just saw I screwed up the mention in my comment on Sunday. I moved and renamed the constants to fit proper standards. Do you think this change is good like that or are there more things to improve?

Regards,
Julian

bdraco reviewed (approved):
This is a good first step.

In a followup, it would be great if the direct call to `rest.RestSwitch` was eliminated and everything we tested after calling `assert_setup_component` and using service calls + checks on the state machine.
